### PR TITLE
pytest framework and CI

### LINKED ISSUE #2: add testing with pytest
Add testing with [pytest](https://docs.pytest.org/en/7.3.x/)
- [x] create `tests` directory under `src/sparsetrace` (see #1 for reorganizing as a package(
- [x] start with initial `test_directintersection.py` and move the [testing code from directintersection.py](https://github.com/jcandane/sTr/blob/f15d06083bceeff429d0f300e69834fe0d21d33f/src/directintersection.py#L107C1-L116C85) to a new file.

## running tests
Once you have your first test written, you should be able to do (from the top directory)
```bash
pytest -v
```
and see the tests being executed.

## test_directintersection.py
The test would roughly look like the following
```python
import pytest
import torch

from ..uniform_random_sparse_tensor import uniform_random_sparse_tensor
# note: we will later change relative imports to imports from
# an installed package:
## from sparsetrace.uniform_random_sparse_tensor import uniform_random_sparse_tensor

from ..directintersection import twotrace

# re-usable testing data
@pytest.fixture
def A():
   return uniform_random_sparse_tensor(35, torch.tensor([12,4,8,3]))

@pytest.fixture
def B():
   return uniform_random_sparse_tensor(45, torch.tensor([6,12,8]))

@pytest.fixture
def adjmatrix():
   return torch.tensor([[0,1],[2,2]]).T

@pytest.fixture
def reference(A, B, adjmatrix):
   ## ideally, use torch or numpy ops for ref (i.e., *independent* calculation);
   ## for right now, this works as a regression test, not necessarily as a 
   ## correctness test
   return twotrace(A.to_dense(), B.to_dense(), adjmatrix)

def test_twotrace_sparse_sparse(A, B, adjmatrix, reference):
   C = twotrace(A, B, adjmatrix)
   assert torch.allclose(C.to_dense(), reference)

def test_twotrace_dense_sparse((A, B, adjmatrix, reference):
   C = twotrace(A.to_dense(), B, adjmatrix)
   assert torch.allclose(C.to_dense(), reference)

def test_twotrace_sparse_dense((A, B, adjmatrix, reference):
   C = twotrace(A, B.to_dense(), adjmatrix)
   assert torch.allclose(C.to_dense(), reference)
```

### PR DESCRIPTION
**Pull request by orbeckst**
- fix #2 and #8 
- Run tests with
  ```
   pytest
  ```
  or
  ```
   pytest -v
  ```
- run tests via GitHub actions on Linux, macOS, Windows for Python 3.8 - 3.11 (continuous integration)
- coverage is collected and sent to codecov.io (but still requires set up)
- check package build for every commit (continuous integration)

### CODE REVIEW

File: bisum/tests/test_bisum.py [hunk 1]
```diff
@@ -6,98 +6,70 @@
...
+    return torch.rand(10,10,10, device=device)
+
+@pytest.fixture(scope="module")
+def C():
+    return torch.rand(13,13,13, device=device)
+
+
+
+@pytest.mark.parametrize("einsumstr", [
+    "aeacec,cdd -> aed",  #SPARSE-SPARSE, NO post-inter-externals, NO post-transpose
```
Comment (orbeckst) [inline bisum/tests/test_bisum.py:37]: Note that the tests can be parametrized so you can just add another `einsumstr` and it will run as its own tests with no additional coding necessary.

File: bisum/tests/test_bisum.py [hunk 2]
```diff
@@ -6,98 +6,70 @@
...
+])
+def test_sparse_sparse_no_post_inter_externals(A_big, B, einsumstr):
```
Comment (orbeckst) [inline bisum/tests/test_bisum.py:39]: The only difference that I noticed between `test_sparse_sparse_no_post_inter_externals()` and `test_sparse_sparse_with_post_inter_externals()` was the use of a different `A` tensor in the example. It is also possible to parametrize over different fixtures but it's a bit more involved and I couldn't be bothered to dig into how exactly to do it. But just know that we can just do Cartesian products of tests (e.g. 3 different "A" and 2 different "B" and 4 different "einsumstr" would create 3*2*4 = 24 parametrized tests).

File: bisum/tests/test_bisum.py [hunk 3]
```diff
@@ -6,98 +6,70 @@
...
+    ### cdd -> cd  ----> accd ----> acd
+
+    assert torch.allclose(
+        bisum(einsumstr, A.to_sparse(), B.to_sparse()).to_dense(),
+        torch.einsum(einsumstr, A, B))
+
+
+def test_sparse_sparse_with_empty_array_1(C, B):
+    # if slicing returns an empty array avoid lexsort!!!!!!!!!!!!!!!! AND "dad,mom -> dm"
+    einsumstr = "aaa,wop -> a" ## SPARSE-SPARSE, YES post-inter-externals, YES post-transpose
```
Comment (orbeckst) [inline bisum/tests/test_bisum.py:62]: You can also parametrize this tests if you like. I just kept it as a single test because I didn't know what was special about this one compared to the `test_sparse_dense_with_empty_array_2()` one.

**Comment by orbeckst** [comment]
I'll see if I can quickly set up basic continuous integration so that the tests are run on every commit.

**Comment by orbeckst** [comment]
FYI, after installing `conda install pytest coverage pytest-cov` I ran a coverage test of the current code
```
pytest --cov=bisum
```
and got
```
---------- coverage: platform darwin, python 3.10.0-final-0 ----------
Name                         Stmts   Miss  Cover
------------------------------------------------
bisum/__init__.py                0      0   100%
bisum/bisum.py                  43     17    60%
bisum/dense_intra.py            35     27    23%
bisum/generic_functions.py      62     43    31%
bisum/labels.py                 65     58    11%
bisum/sparse_intra.py          158    136    14%
bisum/tensor_to_matrix.py      102     94     8%
bisum/tensordot.py             126    103    18%
bisum/tests/test_bisum.py       28      0   100%
------------------------------------------------
TOTAL                          619    478    23%
```
which indicates that the current tests only touch a small fraction (23%) of the current code. It would be good to write more tests that exercise other portions of the code. 

When we have CI running, you should also set up an account at https://codecov.io so that we can easily see which lines of code are not covered by tests.

**Comment by orbeckst** [comment]
You might also want to look into using [hypothesis](https://hypothesis.readthedocs.io/en/latest/) for testing as it can generate tests itself, using constraints. It's good at finding edge cases.

**Comment by orbeckst** [comment]
Congratulations @jcandane , your code runs on Linux, macOS, and Windows — at least the initial tests all pass. On Linux, Python 3.8 - 3.11 is confirmed to work (on macOS and Windows we are only testing 3.10 because these GitHub actions runners take longer).

**Comment by orbeckst** [comment]
Once this PR is merged, you can also set up GitHub in such a way (via repository settings), that tests have to pass before a PR can get merged. In this way it becomes a lot easier to handle contributions from external developers.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+[build-system]
+requires = ["setuptools>=40.8.0", "wheel"]
+
+[project]
+name = 'bisum'
+version = '0.1.1'
```
Comment (orbeckst) [inline pyproject.toml:6]: Right now, the version must be updated in the pyproject.toml _and_ the setup.py file. 

I think we can just get rid of setup.py but we need to see how to add the README/long description in pyproject.toml. That can be a different issue/PR.

Comment (orbeckst) [reply pyproject.toml:7]: Just removed setup.py, so only pyproject.toml remains relevant.

**Review by orbeckst — COMMENTED** [review]
I am using this PR for an overall modernization
- pytest
- GitHub actions for testing and coverage
- update build system to modern pyproject.toml

**Review by jcandane — APPROVED** [review]
